Add documentation about tabtests (#106)
xodio/xod-docs: docs/guide/testing-patches/README.md

Also, tests often serve as excellent documentation of what is expected at outputs with specific parameters at inputs.

Testing nodes in XOD implemented as tabular tests (tabtests for short). That means a test suite looks like a table. The first row contains pin labels of a tested node. All the next rows are test cases, one row is the one test case and one [transaction](/docs/guide/execution-model/) of the XOD program. In columns, input values and expected outputs for corresponding pin labels.

<div class="ui segment">
<span class="ui ribbon label">Input values and expected outputs</span>
Input values and expected outputs defined by tabtest literals, which generally is the same as literals in Inspector, but have some special literals.
See the full list of tabtest literals in the <a href="/docs/reference/tabtests/">reference</a>.
</div>

When you run tests, they either pass or fail with an error describing the difference between actual and expected outputs.

However, there is one limitation, you cannot test nodes, that work with hardware modules.

> ## Testing nodes with primitive types

Reviewer: ```suggestion
## Primitive types
```

The context is known. Keep the headings short

Nodes with such primitive types as Strings, Numbers, Booleans are simplest to test. Let's imagine that there is no `xod/core/to-percent` node in the XOD standard library, and we implement it by ourselves.

![to-percent patch contents](./to-percent.patch.png)

It's a simple node that converts some number into a string with the concatenated percent character. However, what's in the output if it takes a `NaN` on the input, or `-Inf`?

First of all, place the new marker node: `xod/patch-nodes/tabtest`.

> Now you can double-click on `tabtest` node, and an embedded tabtest editor opens up. It looks like a table, let's write test cases:

Reviewer: ```suggestion
Now double-click the `tabtest` node. An embedded tabtest editor opens up which looks like a table. Let's write test cases:
```

Also, suggest attaching this paragraph to the previous one.

<table class="ui table compact">
  <thead>
    <tr>
      <th>NUM</th>
      <th>PERC</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>0</td>
      <td>"0%"</td>
    </tr>
    <tr>
      <td>1</td>
      <td>"100%"</td>
    </tr>
    <tr>
      <td>25.5</td>
      <td>"2550%"</td>
    </tr>
    <tr>
      <td colspan="2">// Test rounding</td>
    </tr>
    <tr>
      <td>0.735</td>
      <td>"74%"</td>
    </tr>
    <tr>
      <td>0.734</td>
      <td>"73%"</td>
    </tr>
    <tr>
      <td colspan="2">// Negative values</td>
    </tr>
    <tr>
      <td>-0.6</td>
      <td>"-60%"</td>
    </tr>
    <tr>
      <td colspan="2">// Special values</td>
    </tr>
    <tr>
      <td>NaN</td>
      <td>"NaN"</td>
    </tr>
    <tr>
      <td>Inf</td>
      <td>"Inf"</td>
    </tr>
    <tr>
      <td>-Inf</td>
      <td>"-Inf"</td>
    </tr>
    <tr>
      <td>9999999999999</td>
      <td>"OVF"</td>
    </tr>
  </tbody>
</table>

<div class="ui segment">
<span class="ui ribbon label">Note</span>

> If you save the project as a multi-file project, you can find out `patch.test.tsv` file in the directory of your patch. So you can write it manually if you like. Pay attention that columns are separated by tabs (`\t`).

Reviewer: Put empty lines to have Markdown parsed properly

</div>

> All right, we just wrote tests. It's time to run them.

Reviewer: ```suggestion
All right, we just wrote the tests. It's time to run them.
```

> Press the "Run tests" button in the panel above and wait for a little. It can take some time because these tests are transpiling to the C++ code, then compile in the cloud and then it executes. You can watch for the progress in the Deployment pane below: you'll see a progress bar, and you can show a pane to see details.

Reviewer: ```suggestion
Press the "Run tests" button in the panel above and wait for a little. It can take a few seconds because these tests are transpiling to the C++ code, then compile in the cloud, and then execute. You can watch for the progress in the Deployment pane below: you'll see a progress bar. Expand the pane to see more details.
```

> Tests failed! Let's inspect details in the "Tests" tab in deployment pane. Here is the lines that makes sense to us:

Reviewer: ```suggestion
Tests failed! Let's inspect details in the "Tests" tab in deployment pane. Here are the lines that are important to us:
```

```
    ./to-percent.catch.inl:36: FAILED:
      REQUIRE( probe_PERC.state.lastValue == xod::XStringCString("NaN") )
    with expansion:
      "NaN%" == "NaN"
```

It tells us that we get value `"NaN%"` in the output instead of expected `"NaN"`. We just caught a bug! Let's fix it and rerun tests:

> ![Fixed to-percent patch contents](./to-percent-fixed.patch.png)

Reviewer: I’m not sure it is a bug. What’s important we get an over-complicated patch in exchange for nothing. Maybe it’s better to make an intentional mistake in the expectation and then fix it. Something like rounding mistake or letter casing mistake.

Author: Okay... you're right.
I just want to show how helpful tests can be 😬
Probably, I have to find a better example... 🤔

Now you see a green notification message about passed tests.

You just wrote tests, found a bug and fixed it. Congratulations!

<div class="ui segment">
<span class="ui ribbon label">Note</span>

> You can write tabtests for the nodes with C++ implementation in the same way, just place `xod/patch-nodes/tabtest` marker node and fulfill the table.

Reviewer: ```suggestion
You can write tabtests for the nodes with C++ implementation in the same way, just place `xod/patch-nodes/tabtest` marker node and fill the table.
```

Empty lines are missing

</div>

<a name="impureness"></a>

> ## Impure tests for impure nodes

Reviewer: ```suggestion
## State keeping
```

?

And you don’t have to put an alternative anchor then.

> We just tested the pure node. But how to deal with impure, like `xod/core/count`?

Reviewer: ```suggestion
We just tested a pure node. That is a node which depends only on its input values when producing output. But how to deal with impure nodes, like `count`, which also depend on the internal state and past?
```

![count patch contents](./count.patch.png)

> All right, we did not tighten the nuts so that we can test these nodes too, but we have to keep it in mind when we are writing tests for them. Each next test case did not reset the state of the previous test cases, so if you send pulse twice in the `INC` pin of the `count` node, it outputs `2` in the second case. If you want to start over — add a new test case and send `pulse` to the `RST` pin.

Reviewer: ```suggestion
All right, we did not tighten the nuts so that we can test these nodes too, but we have to keep it in mind when we are writing tests for them. Each subsequence test case does not reset the state of the previous test case, so if you send a pulse to the `INC` pin of the `count` node twice, it outputs `2` in the second case. If you want to start over — add a new test case and send `pulse` to the `RST` pin.
```

<table class="ui table compact">
  <thead>
    <tr>
      <th>STEP</th>
      <th>INC</th>
      <th>RST</th>
      <th>OUT</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>1</td>
      <td>no-pulse</td>
      <td>no-pulse</td>
      <td>0</td>
    </tr>
    <tr>
      <td>1</td>
      <td>pulse</td>
      <td>no-pulse</td>
      <td>1</td>
    </tr>
    <tr>
      <td>1</td>
      <td>pulse</td>
      <td>no-pulse</td>
      <td>2</td>
    </tr>
    <tr>
      <td>1</td>
      <td>no-pulse</td>
      <td>pulse</td>
      <td>0</td>
    </tr>
    <tr>
      <td colspan="4">// And so on...</td>
    </tr>
  </tbody>
</table>

<div class="ui segment">
<span class="ui ribbon label">Note</span>

> Do not forget to test edge cases like pulses in both pulse pins `INC` and `RST`, inject `NaN` or `Inf` as steps in counter and so on. It helps to catch bugs, make the behavior of the node more obvious, and provide better documentation.

Reviewer: Empty lines

</div>

## Time flow

> Some impure nodes are dependent on a time flow, like `xod/core/fade` or `xod/core/delay` and etc. To test it we can mock the time.

Reviewer: ```suggestion
Some impure nodes are dependent on the time flow. For example, `fade` and `delay`. To test them we can mock the time.
```

> Just specify the column with unique magic name `__time(ms)` and specify the time in milliseconds in this column. It supports integers only. So you can travel through time and even go back in time.

Reviewer: ```suggestion
Make a column with the magic name `__time(ms)` and specify timestamps in milliseconds in this column. Such way you can travel through time.
```

However, also you have to know two things:

> - Nodes that use `isTimedout` function inside or contains nodes like `xod/core/delay`, `xod/core/clock`, are marked as timed out only at the next millisecond. For example, if we set a delay for 500ms, it pulses `DONE` on the 501ms.

Reviewer: Add:

> - A timestamp should be a non-negative integer.

> - Nodes that use `isTimedout` function inside or contains nodes like `xod/core/delay`, `xod/core/clock`, are marked as timed out only at the next millisecond. For example, if we set a delay for 500ms, it pulses `DONE` on the 501ms.

Reviewer: ```suggestion
- Nodes that use `isTimedout` function inside or contain nodes like `xod/core/delay`, `xod/core/clock`, are marked as timed out only at the next millisecond. For example, if we set a delay for 500 ms, it pulses `DONE` on the 501 ms.
```

> - When you do not use this special column, we automatically increment time flow in the tests by one millisecond before running each test case. It provides correct working of `defer` nodes inside. However, when you use `__time(ms)` column you have to update it manually. So if you want to test something with the same time at each row, you probably do not need this column at all.

Reviewer: ```suggestion
- When you do not use this special column, the framework automatically increments the time value by one millisecond before running each test case. It is necessary to correctly support `defer` nodes. However, when you use the `__time(ms)` column you have to update it manually. So if you want to test something with at the same time point at each row, you probably do not need this column at all.
```

You can checkout tabtests in these nodes to understand how it looks and works: `xod/core/delay`, `xod/core/clock`, `xod/core/fade`, `xod/core/system-time`.

> 

Reviewer: An example from either of them would be nice

<div class="ui segment">
<span class="ui ribbon label">Note</span>

> We recommend placing `__time(ms)` column at first place to keep it clear and see what time is it. However, you can place it in any place of the table.

Reviewer: Empty lines

</div>

## Variadics, generics, custom types

> To test nodes with variadics, generics or custom types do the next:

Reviewer: ```suggestion
To test nodes with variadics, generics or custom types do the following:
```

Hyperlinks to variadics and generics would be nice

> 1. Create a special patch with name describes the test and put your node or nodes that you want to cover with test inside

Reviewer: ```suggestion
1. Create a special patch with a name describing the test and put your node(s) to cover with test inside
```

> 2. Place terminals with primitive types

Reviewer: I would link “Place terminals” to the corresponding section of “Making own patch nodes”

      <td>no-pulse</td>
      <td>0</td>
      <td>no-pulse</td>
    </tr>
    <tr>
      <td>1</td>
      <td>pulse</td>
      <td>2</td>
      <td>no-pulse</td>
      <td>3</td>
      <td>no-pulse</td>
      <td>4</td>
      <td>no-pulse</td>
      <td>1</td>
      <td>pulse</td>
    </tr>
    <tr>
      <td>1</td>
      <td>no-pulse</td>
      <td>2</td>
      <td>pulse</td>
      <td>3</td>
      <td>no-pulse</td>
      <td>4</td>
      <td>no-pulse</td>
      <td>2</td>
      <td>pulse</td>
    </tr>
    <tr>
      <td>1</td>
      <td>no-pulse</td>
      <td>2</td>
      <td>no-pulse</td>
      <td>3</td>
      <td>pulse</td>
      <td>4</td>
      <td>no-pulse</td>
      <td>3</td>
      <td>pulse</td>
    </tr>
    <tr>
      <td>1</td>
      <td>no-pulse</td>
      <td>2</td>
      <td>no-pulse</td>
      <td>3</td>
      <td>no-pulse</td>
      <td>4</td>
      <td>pulse</td>
      <td>4</td>
      <td>pulse</td>
    </tr>
    <tr>
      <td>1</td>
      <td>no-pulse</td>
      <td>2</td>
      <td>no-pulse</td>
      <td>3</td>
      <td>no-pulse</td>
      <td>4</td>
      <td>no-pulse</td>
      <td>0</td>
      <td>no-pulse</td>
    </tr>
  </tbody>
</table>

---

> Well, now you know how to write and run tests. So now it might help you to write more stable programs.

Reviewer: ```suggestion
Now you know how to write and run tests. It will help you to write more stable programs.
```

> Moreover, this is the time to tell you about test-driven development. It's a development process when a developer writes tests before write code. It provides productivity, quality, and confidence to the final program. Also, when you write test cases, some things can become evident for you, like "it should be two small nodes instead of the huge one" or "oh, it could return invalid values and how I should deal with it?"

Reviewer: ```suggestion
Moreover, this is the time to tell you about test-driven development. It's a development process when a developer writes tests before writing code. It brings productivity, quality, and confidence to the development process. Also, when you write test cases, some things can become evident for you, like "it should be two small nodes instead of the huge one" or "oh, it can return invalid values and how I should deal with it?"
```

Nice paragraph. I suggest elaborating and adding a link to https://en.wikipedia.org/wiki/Test-driven_development. For example:

> ...about test-driven development (known as \[TDD\](the-link))...

> Also, its a lot faster to run tests than upload code to the board after each change and test all possible cases manually.

Reviewer: ```suggestion
Also, it is a lot faster to run tests than upload code to the board after each change and test all possible cases manually.
```

> So we recommend using TDD when you're making some nodes that are not about hardware and detach all logic from the hardware to test it quickly.

Reviewer: ```suggestion
So we recommend using TDD when you're making some nodes that are not about hardware. Detach all the logic from the hardware to test it quickly.
```

Let's summarize:

1. Tests help to make a stable program.
2. Tests serve as proper documentation.

> 3. Tests provide developer improving the code without worrying about code consumers.

Reviewer: ```suggestion
3. Tests help you improve the program without worrying to break things.
```
